semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Main.scala

```diff
@@ -35,32 +45,46 @@ object Main {
       failed = true
     }
     val classpath = Classpath(settings.cps.mkString(File.pathSeparator))
+    val scopes = new TypeVariableScopes()
     classpath.visit { root =>
       new FileVisitor[Path] {
         // Convert a .class file to a .class.semanticdb file with symbols only.
         override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
+          FileVisitResult.CONTINUE
+        }
+
+        def handleFile(file: Path): FileVisitResult = {
           if (PathIO.extension(file) != "class") return FileVisitResult.CONTINUE
           try {
             val relpath = AbsolutePath(file).toRelative(root).toString
-            val bytecode = ByteCode(Files.readAllBytes(file))
+            val bytes = Files.readAllBytes(file)
+            val bytecode = ByteCode(bytes)
             val classfile = ClassFileParser.parse(bytecode)
-            ScalaSigParser.parse(classfile) match {
-              case Some(scalaSig) =>
-                val className = NameTransformer.decode(PathIO.toUnix(relpath))
-                val semanticdbRelpath = relpath + ".semanticdb"
-                val semanticdbAbspath = semanticdbRoot.resolve(semanticdbRelpath)
-                val semanticdbDocument = s.TextDocument(
-                  schema = s.Schema.SEMANTICDB3,
-                  uri = className,
-                  language = Some(s.Language("Scala")),
-                  symbols = scalaSigPackages(scalaSig) ++ scalaSigSymbols(scalaSig))
-                val semanticdbDocuments = s.TextDocuments(List(semanticdbDocument))
-                FileIO.write(semanticdbAbspath, semanticdbDocuments)
-              case None =>
-                // NOTE: If a classfile doesn't contain ScalaSignature,
-                // we skip it for the time being. In the future, we may add support
-                // for parsing arbitrary Java classfiles.
-                ()
+
+            val isScalaFile =
+              classfile.attribute("ScalaSig").isDefined ||
+                classfile.attribute("ScalaInlineInfo").isDefined
+            val language = if (isScalaFile) "Scala" else "Java"
+            val semanticdbInfos: Option[Seq[s.SymbolInformation]] = if (isScalaFile) {
+              ScalaSigParser.parse(classfile).map { scalaSig =>
+                val semanticdbInfos = scalaSigPackages(scalaSig) ++ scalaSigSymbols(scalaSig)
+                semanticdbInfos
+              }
+            } else {
+              val infos = Javacp.sdocument(root.toNIO, file, scopes).symbols
```

Comment:
I think that the fact that `Javacp` constructs a document and then its only usage is to deconstruct it suggests that the interface between `Main`, `Javacp` and (the virtual) `Scalacp` needs more work.

I suggest that we:  
  1) Factor out `Scalacp` into a separate file.
  2) Unify the contract of `Scalacp` and `Javacp` to say `def sinfos(node: ClassNode): List[s.SymbolInformation]` with other methods being private.
  3) Have `Main` handle assembling those infos into documents.

I can do this refactoring after the pull request is merged.